### PR TITLE
Push filtered bytes instead of concat for 100x better performance

### PR DESCRIPTION
**Pull request by SukantPal**
Hey, greetings from [Teamflow](https://www.teamflowhq.com)!

# Change

TLDR; Using `push()` instead of `concat()` saves lots of CPU time when the state is large.

# Context

We started using `@filter`/`@filterChildren` and noticed a very steep drop in server performance when the state was large. When testing with ~900 furniture objects with ~10 fields per object, `applyFilters` took 6 seconds on my 3.6 GHz Core i9 processor:

screen_shot_2021-05-26_at_3 48 06_pm

When I dug into this, found that Array concatenation itself was taking 4.6 seconds:

screen_shot_2021-05-26_at_5 28 27_pm

Using concatenation causes excessive copying of larger and larger arrays as the loop progresses, making it an O(n²) operation instead of O(n). By pushing the bytes into the same array, it performance 100x faster and finishes in less than 60ms:

screen_shot_2021-05-26_at_5 48 20_pm

### CODE REVIEW

File: src/Schema.ts [hunk 1]
```diff
@@ -806,15 +806,15 @@ export abstract class Schema {
                         //
                         // use cached bytes directly if is from Schema type.
                         //
-                        filteredBytes = filteredBytes.concat(changeTree.caches[fieldIndex]);
+                        filteredBytes.push.apply(filteredBytes, changeTree.caches[fieldIndex] || [undefined]);
```
Comment (foolmoron) [inline src/Schema.ts:809]:
```suggestion
                        filteredBytes.push(...(changeTree.caches[fieldIndex] ?? []));
```

File: src/Schema.ts [hunk 2]
```diff
@@ -806,15 +806,15 @@ export abstract class Schema {
...
                         containerIndexes.add(fieldIndex);
 
                     } else {
                         if (containerIndexes.has(fieldIndex)) {
                             //
                             // use cached bytes if already has the field
                             //
-                            filteredBytes = filteredBytes.concat(changeTree.caches[fieldIndex]);
+                            filteredBytes.push.apply(filteredBytes, changeTree.caches[fieldIndex] || [undefined]);
```
Comment (foolmoron) [inline src/Schema.ts:817]:
```suggestion
                            filteredBytes.push(...(changeTree.caches[fieldIndex] ?? []));
```

**Review by foolmoron — CHANGES_REQUESTED** [review]
This is sick! Filtering is basically unusable without this.

Left some suggestions to use `...` (slightest bit faster than `apply`) and avoid pushing an `undefined` into the array in the default case.

**Comment by Elyx0** [comment]
Good catch 😉 Glad you're still using Coly and giving back.

File: src/Schema.ts (hunk 1 above)
Comment (SukantPal) [reply src/Schema.ts:809]: You have to push `undefined` into the array otherwise unit tests fail.

**Comment by endel** [comment]
Nice catch indeed, thanks for your PR @SukantPal! I do think @foolmoron is relevant, perhaps the failing test has an issue (just noticed it's calling `state.encodeAll()` without `useFilters=true` as argument). Gonna review this soonish!

**Comment by endel** [comment]
Thanks a lot for the PR @SukantPal, and @foolmoron for the suggestion! 🎉 

Just released version `1.0.24` with these changes!

The test that was not passing actually had a problem (not using encodeAll+useFilters), I've updated it and it's passing again!
